ENH: Add 3D geometry and propagation module

mdw771 opened the pull request:
Geometry:
- Added `bounding_box` and `contains` methods in all 3D geometry classes, `Cuboid_3d`, `Sphere_3d`, `TruncatedCone_3d`, and `Cylinder_3d`. 

Material:
- Added `CustomMaterial` class which takes user specified delta and beta. Useful for describing mixtures. 

Plot:
- `discete_grometry` now works with 3D geometry classes. 
- Multiple changes made to `discrete_phantom`:
  - The `prop` argument can take a list of strings (e.g. ['delta', 'beta']) to generate grids for multiple quantities. Result grids are returned in list if a list or tuple object is passed. 
  - Added `energy` keyword argument. 
  - Added `fix_psize` argument. When set to `False`, the function behaves as what it originally is. When set to `True`, pixel size is fixed at 1 instead of being given by `1.0 / size`. This is useful in preventing a over-fine meshgrid in 3D cases which results in intractible computational cost. 
  - Added `overlay_mode` argument. When set to `add`, the function behaves as what it originally is. When set to `replace`, geometries in child phantoms replace the voxels with parent values, instead of adding to them. 
  - Added `return_patch` argument. When set to `True`, the function also returns the geometry patch (as `bool` type). 
  - A `try except` structure is used to allow the reading of properties in form of both methods and attributes (such as `delta` and `beta` in `CustomMaterial`). 

Acquisition:
- Added `Simulator` class. The object takes in refractive index grids and performs multislice propagation. This class should be used in place of `initialize_wavefront` and `multislice_propagation` in the previous `propagation` module.

Demo codes:
- Amended `tests/beyond_dof.py` to provide a demo for functions added. 

Other changes:
- Removed `grid`.

carterbox commented:
@mdw771, Please fix the error that cause failing Travis builds. It looks like you forgot to remove the import statements for the `grid` module. There is also an import error for `skimage.morphology` probably because `skimage` isn't listed in the `requirements.txt` or in the install section of `travis.yml`.

We are targeting both `python2.7` and `python3.x`, so you need to use parenthesis on print statements.

Thanks for your patience in advance. I'm going on vacation next week, but I'll try and review as much as I can today.

carterbox commented on tests/3d_brain.py:
```diff
@@ -0,0 +1,87 @@
+    save_pack = np.append(save_pack, np.array([energy]))
+    np.savetxt('sav/dim.txt', save_pack)
+
+print 'Slice check.'
```
print needs parenthesis for `python3.x`

carterbox commented on tests/3d_brain.py:
```diff
@@ -0,0 +1,87 @@
+from xdesign.grid import *
```
We use [`nosetests`](http://nose.readthedocs.io/en/latest/) for our test scripts so that they can be run automatically on `Travis` and so that debug logging is automatically set when they are run. The quick and dirty is that all tests should be in files with the word `test` in the filename and each test should be a function within that file with `test` in the title.

carterbox commented on xdesign/__init__.py:
```diff
@@ -56,6 +56,8 @@
 from xdesign.plot import *
 from xdesign.material import *
+from xdesign.propagation import *
+from xdesign.grid import *
```
Changelog says you removed the grid module, but here is an import statement for it. Ooops!

carterbox commented on xdesign/acquisition.py:
```diff
@@ -63,12 +63,21 @@
+try:
+    import queue
+except:
+    import Queue as queue
```
Thanks for patching this; I forgot that this was a bug.

carterbox commented on xdesign/acquisition.py:
```diff
@@ -540,3 +549,183 @@ def raster_scan(sx, sy):
+    energy : float
+        Beam energy in keV. Should match the energy used for creating the grids.
+    psize : list
+        Pixel size in nm.
```
All parameters should be expressed in centimeter gram seconds electron volts.

carterbox commented on xdesign/acquisition.py:
```diff
@@ -540,3 +549,183 @@ def raster_scan(sx, sy):
+        Parameters:
+        -----------
+        type : str
+            Type of wavefront to be initialized.
```
Please enumerate the options of `type` and `kwargs` in the documentation because readers may not be looking at the source code.

carterbox commented on xdesign/acquisition.py:
```diff
@@ -540,3 +549,183 @@ def raster_scan(sx, sy):
+        elif part == 'beta':
+            tifffile.imshow(self.grid_beta)
+        else:
+            print('WARNING: wrong part specified for show_grid.')
```
Use the [`warnings`](https://docs.python.org/3/library/warnings.html) module for warnings.

carterbox commented on xdesign/acquisition.py:
```diff
@@ -540,3 +549,183 @@ def raster_scan(sx, sy):
+def tomography_3d(grid, wavefront, probe, ang_start, ang_end, ang_step=None, n_ang=None, free_prop_dist=None, savefolder='tomo_output', fname='tomo', format='h5', pr=None, **kwargs):
+    allowed_kwargs = {'mba': ['alpha']}
+    if pr not in allowed_kwargs:
+        raise ValueError
```
Please include helpful messages with errors.

carterbox commented on xdesign/acquisition.py:
```diff
@@ -540,3 +549,183 @@ def raster_scan(sx, sy):
+        angles = np.linspace(ang_start, ang_end, n_ang)
+        ang_step = float(ang_end - ang_start) / (n_ang - 1)
+    else:
+        print('ERROR:xdesign.acquisition:Angular step or number of angles should be specified.')
```
Use an error instead of print.

carterbox commented on xdesign/acquisition.py:
```diff
@@ -540,3 +549,183 @@ def raster_scan(sx, sy):
         p.translate(-1)
         p.rotate(theta, Point([0.5, 0.5]))
+
+def tomography_3d(grid, wavefront, probe, ang_start, ang_end, ang_step=None, n_ang=None, free_prop_dist=None, savefolder='tomo_output', fname='tomo', format='h5', pr=None, **kwargs):
```
This function needs a docstring.

carterbox commented on xdesign/geometry.py:
```diff
@@ -1156,6 +1164,191 @@ def __init__(self, center, side_length=None, radius=None):
+        if not isinstance(top_center, Point):
+            raise TypeError("center must be of type Point.")
+        super(Cylinder_3d, self).__init__(top_center, length, radius, radius)
+
```
Cylinder is missing a `contains` method that is different from the cone.

carterbox commented on xdesign/metrics.py:
```diff
@@ -61,7 +61,7 @@
 from scipy import optimize
 from scipy.stats import norm, exponnorm, expon, ttest_ind
-from phasepack import phasecongmono as _phasecongmono
+#from phasepack import phasecongmono as _phasecongmono
```
This will cause the VIFp metric to fail, I believe.

carterbox commented on xdesign/plot.py:
```diff
@@ -367,26 +377,54 @@ def discrete_phantom(phantom, size, ratio=9, uniform=True,
+        if fix_psize:
+            psize = 1
+        else:
+            psize = 1.0 / size
```
I don't understand the purpose of the new `fix_psize` parameter. Isn't it equivalent to setting the `size` parameter to `1`? We could also just change the interface to set `psize` directly instead of using `size`. `size` is just a holdover from when I initially designed this function for the `plot` module. It was more convenient to say "I want a 256x256 grid" instead of "I want the pixel size to be 1/256 cm".

carterbox commented on xdesign/plot.py:
```diff
@@ -367,26 +377,54 @@ def discrete_phantom(phantom, size, ratio=9, uniform=True,
+            try:
+                value = getattr(phantom.material, i_prop)(energy)
+            except:
+                value = getattr(phantom.material, i_prop)
```
First, we shouldn't have a bare exception. We should always name the exception we intend to catch. I think in this case we are looking for `TypeError` because the energy argument will be rejected if the attribute isn't callable.

Related: what if we just change all material properties to this signature:
```python
def some_material_property(self, **kwargs):
    energy = kwargs['energy']
```
Then an error is only raised if you're trying to get a property without enough info, but if you have extra info, it doesn't matter.

carterbox commented on xdesign/propagation.py:
```diff
@@ -0,0 +1,234 @@
+logger = logging.getLogger(__name__)
+
+
+__author__ = "Daniel Ching, Doga Gursoy"
```
You should change the author string to your name because Doga and I haven't contributed to this module.

carterbox commented on xdesign/propagation.py:
```diff
@@ -0,0 +1,234 @@
+    return f2
+
+
+def plot_wavefront(wavefront, grid, save_folder='simulation', fname='exiting_wave'):
```
Plot functions should all be in the `plot` module in order to keep import statements minimized.

mdw771 commented:
Thanks for the comments. 
- All occurrences of `import xdesign.grid` have been removed. 
- `tests/3d_brain.py` and `tests/test.py` are outdated scripts used for testing the old version of the package. They have been removed. 
- `tests/beyond_dof.py` has been renamed as `tests/test_tube_particles.py` and the codes are included in function `test_model_prop_pipeline.py`. 
- Units of Input parameters of `Simulator` class were changed to cm and eV. 
- Added docstring to `Simulator.initialize_wavefront`. Options will be explained in documentation in later commits. 
- `warnings` module used for WARNING strings.
- Modified `acquisition.tomography_3d`. Changed phase retrieval method to paganin. `tomopy.retrieve_phase` is used.
- Added `contains` method to `Cylinder_3d`. 
- Added back `import phasepack` in `metrics`.
- I temporarily hold the parameter `fix_psize` in `discrete_phantom` in place. The reason for this extra parameter is that in `discrete_geometry` a grid with side length of approximately `(xmax - xmin) / psize` pixels is created. For 3D object a grid size like this is computationally intractable. `psize` and thus `size` has to be set to 1 to prevent the creation of an over-large grid, but this results in a one-pixel data grid. I suggest that the `discrete` functions takes only the number of pixels per side for the grid and simply assume the real-size of each pixel to be 1, since are not important unless physics calculations (eg. propagation simulation) is involved. 
- `discrete_phantom`: `TypeError` will be raised when invalid material properties are enquired. Property retrieval methods in `Material` and its child classes are modified with keyword argument for energy. 
- Added my name to `__author__` in `propagation` and `geometry`.
- Moved `plot_wavefront` from `propagation` to `plot`.

carterbox commented on xdesign/geometry.py:
```diff
@@ -205,6 +210,9 @@ def midpoint(self, other):
         return self.distance(other) / 2.
 
+    def generate(self, grid, material):
+        pass
```
What is this method for?

mdw771 replied:
Just a placeholder in the parent class. Will remove in the new commit.

carterbox replied:
It's fine if you want to add a placeholder, but it needs a docstring to tell what it is supposed to be placeholding.

carterbox commented:
> I suggest that the discrete functions takes only the number of pixels per side for the grid and simply assume the real-size of each pixel to be 1, since are not important unless physics calculations (eg. propagation simulation) is involved.

I don't understand. Why is the real size of each pixel not important? Every `Entity` has a real size in centimeters, so how is the real size of each pixel not important?

mdw771 commented on tests/test_tube_particles.py:
```diff
@@ -0,0 +1,85 @@
+    try:
+        grid_delta = np.load('data/sav/grid/grid_delta.npy')
+        grid_beta = np.load('data/sav/grid/grid_beta.npy')
+    except:
```
The structure is just a alternative to an `if` judgement: skip modeling if data files exist, otherwise create model and generate grids.

carterbox replied:
Yes, but [`np.load()`](https://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.load.htmll) raises two types of errors. You only want to catch `IOError`. You don't want to catch `ValueError`.

mdw771 commented:
> I suggest that the discrete functions takes only the number of pixels per side for the grid and simply assume the real-size of each pixel to be 1, since are not important unless physics calculations (eg. propagation simulation) is involved.

I just realized that the point coordinates are supposed to be given in cm instead of in pixels. Now this makes more sense to me. 
But I still suggest to change the current scheme that fixes the entire sample size at 1 cm to a scheme that requires inputs of individual pixel size and number of pixels (grid size). When the model is used for microtomography or holographical studies people are generally interested in resolution of micron or nanometer level. It is necessary in such cases to let users specify the pixel size in order to limit the grid size.

carterbox commented:
> people are generally interested in resolution of micron or nanometer level

What I think you mean is: having a 1 cm field of view is too large for some experiments; the ability to specify a smaller region for plotting or discretization is good.

I agree :smile:. I think the best way to handle this is to have two keyword arguments such as below for the discretization function.
```python
"""
psize : float
    The side length of a cubic pixel.
bounding_box : :py:class:`numpy.ndarray`, :py:class:`List`
    The desired field of view specified as `[min, max]` where min and max
    are two column vectors pointing to the min and max corners of the
    desired field of view.
"""
```
Does that sound good?

carterbox commented on xdesign/propagation.py:
```diff
@@ -192,6 +173,7 @@ def far_propagate(grid, wavefront, dist_nm):
 def _far_propagate_2(grid, wavefront, lmd, z_um):
     """Free space propagation using product Fourier algorithm.
     """
+    raise DeprecationWarning
```
If you `raise` a warning instead of using `warnings.warn()`, it will exit the program instead of just displaying the warning.

carterbox commented on xdesign/plot.py:
```diff
@@ -367,26 +377,54 @@ def discrete_phantom(phantom, size, ratio=9, uniform=True,
+            pass
+        return np.squeeze(ret), patch
+    else:
+        return np.squeeze(ret)
```
`ret` will be returned before it is defined if the first `if` statement is false.

This function is supposed to return "0 if phantom has no geometry or material property".

carterbox commented on xdesign/propagation.py:
```diff
@@ -0,0 +1,194 @@
+    delta_nm : float
+        Slice thickness in nm.
+    lmda : float
+        Wavelength in nm.
```
All the measures should be specified in `cgs`.

mdw771 commented:
> I think the best way to handle this is to have two keyword arguments such as below for the discretization function.

This would address the problem, but I'm thinking of another way which seems more straightforward: we set 3 keywords arguments:
```
"""
psize : float
        Side length of a pixel or voxel in cm.
grid_length : float
        Length of a side of the grid in cm (this is the quantity that was originally default to 1 cm).
grid_size : int
        Number of pixels along a side of the grid (this is the original "size" argument).
"""
```
All arguments default to `None` and user shall supply any 2 of them for the function to infer the other one. In this way the grid dimension is not necessarily 1 cm, but the function would be more flexible for various situations. 
How do you think of this?

carterbox commented:
I have two arguments for the 2 arg API over the 2 of 3 arg API.

1. The advantage of specifying the `bounding_box` is that you can reduce computation by avoiding empty regions. `grid_length` and `grid_size` and `psize`, tells **how many** pixels are wanted but **not where** they are wanted, so the problem is under-defined. If you assume the grid includes the origin, this could be expensive if your object of interest is very small and far away from the origin.

2. I oppose the "choose only two of three optional arguments". I tried that with another API, and it made the code messy. You have to check user inputs to make sure only 2 of the 3 choices were supplied, and you have to convert the input to the parameters that the implementation actually uses.

mdw771 commented:
Travis test of `test_tube_particles.test_model_prop_pipeline` fails on server due to the absence of xraylib. After adding `xraylib` in `requirements.txt` and inserting `conda install xraylib` in `.travis.yml`, the building process seems stuck and went timeout after 10 min. Local tests all went through with Python 2.7.13 (Anaconda) except a `Not equal to tolerance rtol=1e-07, atol=0.01` failure in `tests.test_3d.test_probe_circular`.

carterbox commented on xdesign/plot.py:
```diff
@@ -361,32 +376,65 @@ def discrete_phantom(phantom, size, ratio=9, uniform=True,
-        # Make a grid to put store all of the discrete geometries
-        image = np.zeros([size] * phantom.geometry.dim, dtype=float)
+        n_prop = prop.size
+        ret = [None] * n_prop
```
I've implemented this change here: mdw771/xdesign#3 because I think that slicing a single array is more compact than slicing each element of a list separately.

mdw771 commented on xdesign/plot.py:
```diff
@@ -328,17 +328,23 @@ def _make_axis():
 
-def discrete_phantom(phantom, size, ratio=9, uniform=True,
-                     prop='linear_attenuation'):
+def discrete_phantom(phantom, psize, bounding_box=((0, 0), (1, 1)),
```
Yes it was originally because I was trying to keep it consistent with the data format of `geometry.Point`, but I like the idea of column vectors as well. Thanks for the new PR

carterbox commented:
Travis failed because `test_probe_circular`. This test fails like 1/10 times because #53, so I'm ignoring it.
